drivers/dp/swdp_bitbang.c

```diff
@@ -0,0 +1,632 @@
+/*
+ * Copyright (c) 2018-2019 PHYTEC Messtechnik GmbH
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file is based on SW_DP.c from DAPLink Interface Firmware.
+ * Copyright (c) 2009-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+/* Serial Wire Debug Port interface bit-bang driver */
+
+#define DT_DRV_COMPAT zephyr_swdp_gpio
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/swdp.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(swdp, CONFIG_DP_DRIVER_LOG_LEVEL);
+
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+#define CPU_CLOCK		64000000U
+#else
+#define CPU_CLOCK		CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+
+#define CLOCK_DELAY(swclk_freq, port_write_cycles) \
+	((CPU_CLOCK / 2 / swclk_freq) - port_write_cycles)
+
+/*
+ * Default SWCLK frequency in Hz.
+ * sw_clock can be used to overwrite this default value.
+ */
+#define SWDP_DEFAULT_SWCLK_FREQUENCY	1000000U
+
+#define DELAY_SLOW_CYCLES		3U
+
+struct sw_config {
+	struct gpio_dt_spec clk;
+	struct gpio_dt_spec dout;
+	struct gpio_dt_spec din;
+	struct gpio_dt_spec dnoe;
+	struct gpio_dt_spec noe;
+	struct gpio_dt_spec reset;
+	uint32_t port_write_cycles;
+	void *clk_reg;
+	void *dout_reg;
+	void *din_reg;
+	void *dnoe_reg;
+};
+
+struct sw_cfg_data {
+	uint32_t clock_delay;
+	uint8_t turnaround;
+	bool data_phase;
+	bool fast_clock;
+};
+
+static uint8_t sw_request_lut[16] = {0U};
+
+static void mk_sw_request_lut(void)
+{
+	uint32_t parity = 0U;
+
+	for (int request = 0; request < sizeof(sw_request_lut); request++) {
+		parity = request;
+		parity ^= parity >> 2;
+		parity ^= parity >> 1;
+
+		/*
+		 * Move A[3:3], RnW, APnDP bits to their position,
+		 * add start bit, stop bit(6), and park bit.
+		 */
+		sw_request_lut[request] =  BIT(7) | (request << 1) | BIT(0);
+		/* Add parity bit */
+		if (parity & 0x01U) {
+			sw_request_lut[request] |= BIT(5);
+		}
+	}
+
+	LOG_HEXDUMP_DBG(sw_request_lut, sizeof(sw_request_lut), "request lut");
+}
+
+static uint32_t ALWAYS_INLINE sw_get32bit_parity(uint32_t data)
+{
+	data ^= data >> 16;
+	data ^= data >> 8;
+	data ^= data >> 4;
+	data ^= data >> 2;
+	data ^= data >> 1;
+
+	return data & 1U;
+}
+
+static ALWAYS_INLINE void pin_delay_asm(uint32_t delay)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	__asm volatile ("movs r3, %[p]\n"
+			".start_%=:\n"
+			"subs r3, #1\n"
+			"bne .start_%=\n"
+			:
+			: [p] "r" (delay)
+			: "r3", "cc"
+			);
```

Comment:
Wouldn't this work with any Cortex-M? Maybe it could be a `#ifdef CONFIG_CPU_CORTEX_M`

Reply:
@jfischer-no, @fabiobaltieri - I did something similar to this for `fpga_ice40.c`. it can likely be done without resorting to assembly, which makes it slightly more portable.

In the case for the ice40, the DT bindings include a parameter for calibrated delay for exactly a 1 MHz square wave. Maybe I should use the pattern below as well, to allow the frequency to be configurable on a per-system basis.

My suggestions / requests for this:
* an additional DT binding for a calibration table indexed by frequency
  * key: frequency in Hz, val: delay count
  * sorted in increasing order, which would permit both linear and binary search in the driver
  * maybe make that an inline "helper" function in a private header, similar to gpio helpers
* driver API to query the calibration table keys (and therefore min, max, other supported frequencies)
* driver API to query the current frequency
* driver API to set the SWD frequency by frequency instead of count

The reason I'm leaning towards frequency in the API is because it is maybe a bit clearer for users.

Something to consider is whether the `set_frequency` API call would only support exact frequencies or clamping to the nearest frequency. That sort of thing should be consistent between any driver implementation though, so would need to be verified in a testsuite.

Slightly related: https://github.com/zephyrproject-rtos/zephyr/pull/54947#discussion_r1138567407

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/fpga/lattice%2Cice40-fpga.yaml
https://github.com/zephyrproject-rtos/zephyr/blob/bb6fa5707f77d72307008405f19dd83247405392/drivers/fpga/fpga_ice40.c#L132

Reply:
While I agree that it is nice to have the timing parameter provided by device tree, I don't like following:
- The delay is to be calibrated for CPU running the code, not ice40. Note sure if such property should be part of `lattice,ice40-fpga` compatible or rather more related to the core running the code.
- Calibrating for loop that does only handle clock but using it for loop that handles both clock and data. The achieved frequency in actual loop is going to be slower.

Doing a instruction processing based delays is going to be really problematic the more complex CPU this runs on, i.e. the most repeatable results you get on a system without a cache, but the more sophisticated instruction cache or pipeline there is the more annoying it is going to get.

Reply:
did anyone experiment with chained dma descriptors for speed-up?
or even adding an extra input-pin -> IRQ on the swdio, the target ack could be masked on tx, but used as a trigger for ack to speed things up?

nice if it's cpu agnostic of course even though I like the inline asm idea above

seems many run LPC's for their debug probes, these have a Programmable Logic Unit (PLU) feature that could do some simple counting logic

Reply:
> Do we already have an umbrella `zephyr,swdp.yaml`? If not, eventually it might make sense to create one so that `zephyr,swdp-gpio.yaml` and e.g. a PIO implementation could have a common base

That can be done later, I am looking forward #44316